ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PreJoinRepartitionNode.java

```diff
@@ -92,6 +101,11 @@ public LogicalSchema getSchema() {
 
   @Override
   public void setKeyFormat(final FormatInfo format) {
+    // Force repartition in case of schema inference, to avoid misses due to key schema ID mismatch
```

Comment:
That's because it wasn't required that this method be called previously :) I've revamped the logic here so it's now required, as that was the major point of confusion when reading this code as well. Also added the check you suggested.

Reply:
Good call on challenging this, BTW. There was a bug where we weren't always forcing repartitions for upstream joins with SR-enableld key formats in multi-joins. I've fixed the bug and also added multi-join QTTs to cover this.